File: backend/node/api/src/application/use-cases/resume-run.ts
import type { Queue } from "../../application/queue/queue";
import { RunNotFoundError } from "../../domain/errors";
import { inject } from "../../infra/DI/container";
import type { RunRepository } from "../../infra/repository/run";

type Input = {
	runId: string;
};

export class ResumeRunUseCase {
	@inject("run-repository")
	private declare readonly runRepository: RunRepository;

	@inject("queue")
	private declare readonly queue: Queue;

	async execute(input: Input): Promise<[true, undefined] | [undefined, Error]> {
		const [run, error] = await this.runRepository.getById(input.runId);
		if (error) return [undefined, error];
		if (!run) return [undefined, new RunNotFoundError()];

		const [, resumeError] = run.resume();
		if (resumeError) return [undefined, resumeError];

		const [, saveError] = await this.runRepository.update(run);
		if (saveError) return [undefined, saveError];

		await this.queue.publish("run.resume-requested", {
			runId: input.runId,
		});

		return [true, undefined];
	}
}

File: backend/go/api/internal/run/use-cases/resume-run.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/queue"
	runPkg "github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/run"
)

type ResumeRunUseCaseDependencies struct {
	RunRepository *runPkg.RunRepository
	Queue         *queue.RabbitMQ
}

type ResumeRunUseCase struct {
	ResumeRunUseCaseDependencies
}

func NewResumeRunUseCase(deps ResumeRunUseCaseDependencies) ResumeRunUseCase {
	return ResumeRunUseCase{
		ResumeRunUseCaseDependencies: deps,
	}
}

type ResumeRunInput struct {
	RunId string
}

func (u *ResumeRunUseCase) Execute(input ResumeRunInput) error {
	run, err := u.RunRepository.GetById(context.Background(), input.RunId)
	if err != nil {
		return err
	}

	if run == nil {
		return runPkg.RunNotFoundError
	}

	err = run.Resume()
	if err != nil {
		return err
	}

	err = u.RunRepository.Update(context.Background(), run)
	if err != nil {
		return err
	}

	err = u.Queue.Publish("run.resume-requested", runPkg.RunResumeRequestedMapper(run), queue.PublishConfig{})
	if err != nil {
		return err
	}

	return nil
}
